packages/mobx/src/core/atom.ts

```diff
@@ -65,6 +65,9 @@ export class Atom implements IAtom {
      * Invoke this method _after_ this method has changed to signal mobx that all its observers should invalidate.
      */
     public reportChanged() {
+        if (globalState.suppressReportChanged) {
```

Comment:
This feels like a potentially dangerous check, that silently might break reactivity. E.g. do the following cases behave correctly?


```javascript
const x = observable({ a: 1 })
reaction(x => x.a, a => console.log("a", a))
extendObservable(x, { a: 2, b: 1 })
```

```javascript
const x = observable({ a: 1 }, { proxy: true })
// react to new keys
reaction(x => Object.keys(x.a), a => console.log("a", a))

x.b = 1
```

I'm wondering if an alternative approach would be: only increase stateVersion if there are subscribers to the current atom?

Reply:
> This feels like a potentially dangerous check, that silently might break reactivity

Good point. Imo the play here is to make sure it's suppressed only on creation and not on mutation like in this case.
Key changes should invoke reactions and therefore update state version, these two goes hand in hand, a reaction must always see a new state version. Therefore you shouldn't extend observable during render.
Just for clarification: `extendObservable` does notify about key changes.

> only increase stateVersion if there are subscribers to the current atom

Purely from semantic perspective, two different states should aways have different versions, doesn't matter if someone was observing some part of the state or not. That's quite easy to reason about.
Practically speaking, I don't know. I have a bad feeling about it, but I have a hard time to come up with a convincing example. My worry is, that you change some observable `x`, that nobody is observing, so version stays the same. Then something unrelated to mobx - props/state or other external sync store - changes in such a way, that the next render would read from `x`.  Now if react would check whether mobx external state changed, it doesn't know about the change of `x`, even though it's about to use it together with rest of the world (other observables, state, props, other external state) to render. I am not sure if that alone is a problem. The goal here is to avoid inconsistencies and to have an inconsistency, the `x` would probably have to be somehow correlated with the rest of the world (props/state/other store).
But then, if they're correlated, they had to change at the same moment and therefore something, somewhere had to already signal a new version? It's hard for me to contemplate about this, especially when I don't know what react exactly does under the hood and what it's assumptions are. If version always changes, I don't have to think about this...

Reply:
> Purely from semantic perspective, two different states should aways have different versions, doesn't matter if someone was observing some part of the state or not. That's quite easy to reason about

>  If version always changes, I don't have to think about this...

Agreed, I think these principle weighs stronger than my concerns. Let's continue this direction.

Reply:
There is a fundamental problem with subclasses - `new Subclass()` always counts as a state update, because the subclass's constructor calls `makeObservable` on an already observable object (`this`).
Strictly speaking it's correct behavior, because constructors can insert any arbitrary logic between individual `makeObservable` calls and there is no guarantee `new Subclass` is called from transaction, so a reaction invoked in between these `makeObservable` calls should see new state version.

What we would need is to treat all modifications of the same atom inside the same batch it was created as part of initialization. To do that we would need to introduce `globalState.batchId`, assign this `batchId` to every atom when created and suppress `reportChanged` when `atom.batchId === globalState.batchId`. Now you don't have to manually demark initialization sections - they span from atom creation until the end of the batch. If you call `new Subclass()` from action as a good citizen, the state version stays same, no matter how many times you called `makeObservable`. Thoughts?

Reply:
Ok, that sounds smart! It might serve in the future as well to better handle any weird edge cases around reaction creations in actions, recursive updates etc etc, although I believe we handle those all consistently now.